### PR TITLE
port city guide skill

### PR DESCRIPTION
**Pull request by pbheemag**
Building the city guide skill in Java.

## Description
The skill changes include:
* Usage of HttpGet to make external API calls to other services.
* Usage of https://openweathermap.org/ to get weather data.

## Testing
Built a skill to test the changes in this PR.

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist

- [ ] My code follows the code style of this project
- [ ] My change requires a change to the documentation
- [x] I have updated the documentation accordingly
- [x] I have read the **README** document
- [ ] I have added tests to cover my changes
- [ ] All new and existing tests passed

## License

- [x] I confirm that this pull request can be released under the Apache 2 license

[issues]: https://github.com/alexa/alexa-skills-kit-sdk-for-java/issues
[license]: http://aws.amazon.com/apache2.0/
[cla]: http://en.wikipedia.org/wiki/Contributor_License_Agreement

### CODE REVIEW

File: samples/cityguide/models/en-US.json [hunk 1]
```diff
@@ -0,0 +1,98 @@
...
+      "intents": [
+        {
+          "name": "AboutIntent",
+          "samples": [
+            "about",
+            "tell me about this place"
+          ],
+          "slots": []
+        },
+
```
Comment (breedloj) [inline samples/cityguide/models/en-US.json]: Inconsistent line breaks - applies throughout.

File: samples/cityguide/src/com/amazon/ask/cityguide/POJO/Attraction.java [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+
+         http://aws.amazon.com/apache2.0/
+
+     or in the "license" file accompanying this file. This file is distributed on an "AS IS" BASIS,
+     WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for
+     the specific language governing permissions and limitations under the License.
+*/
+
+package com.amazon.ask.cityguide.POJO;
+
```
Comment (breedloj) [inline samples/cityguide/src/com/amazon/ask/cityguide/POJO/Attraction.java:15]: Inconsistent whitespace - applies throughout.

File: samples/cityguide/src/com/amazon/ask/cityguide/POJO/Attraction.java [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+
+public class Attraction {
+
+    private String name;
```
Comment (breedloj) [inline samples/cityguide/src/com/amazon/ask/cityguide/POJO/Attraction.java]: Nit but these should all be final.

File: samples/cityguide/src/com/amazon/ask/cityguide/POJO/CityInformation.java [hunk 4]
```diff
@@ -0,0 +1,102 @@
...
+import java.util.Set;
+
+public class CityInformation {
+
+    private City city;
+    private String state;
+    private Set<Restaurant> restaurants;
+    private Set<Attraction> attractions;
+
+    public CityInformation(City city, String state, Set<Restaurant> restaurants, Set<Attraction> attractions) {
```
Comment (breedloj) [inline samples/cityguide/src/com/amazon/ask/cityguide/POJO/CityInformation.java]: Why a public constructor if you have a builder?

File: samples/cityguide/src/com/amazon/ask/cityguide/handlers/CoffeeIntentHandler.java [hunk 5]
```diff
@@ -0,0 +1,52 @@
...
+public class CoffeeIntentHandler implements IntentRequestHandler {
+    @Override
+    public boolean canHandle(HandlerInput handlerInput, IntentRequest intentRequest) {
+        return intentRequest.getIntent().getName().equals("CoffeeIntent");
+    }
+
+    @Override
+    public Optional<Response> handle(HandlerInput handlerInput, IntentRequest intentRequest) {
+
+        final List<Restaurant> restaurants = SkillUtils.getRestaurantsByMeal(Meal.COFFEE, SkillData.CITY_INFORMATION.getRestaurants());
```
Comment (breedloj) [inline samples/cityguide/src/com/amazon/ask/cityguide/handlers/CoffeeIntentHandler.java]: A lot of this logic seems to be reused for each handler. Is there a way we can generify it to save some code duplication?

File: samples/cityguide/src/com/amazon/ask/cityguide/handlers/GoOutIntentHandler.java [hunk 6]
```diff
@@ -0,0 +1,91 @@
...
+
+    @Override
+    public Optional<Response> handle(HandlerInput handlerInput, IntentRequest intentRequest) {
+        try {
+            JsonNode node = MAPPER.readTree(getWeatherData().getEntity().getContent());
+            final String temperature = node.path("main").path("temp").asText();
+            final String humidity = node.path("main").path("humidity").asText();
+            final String windSpeed = node.path("wind").path("speed").asText();
+
+            final String speechOutput = String.format("Weather in %s is with temperature %s kelvin and %s percent humidity. Wind speed is about %s miles per hour",
```
Comment (breedloj) [inline samples/cityguide/src/com/amazon/ask/cityguide/handlers/GoOutIntentHandler.java]: I don't think I have ever seen the temperature for purposes of a forecast expressed in kelvin.

File: samples/cityguide/src/com/amazon/ask/cityguide/handlers/GoOutIntentHandler.java [hunk 7]
```diff
@@ -0,0 +1,91 @@
...
+        }
+    }
+
+    private CloseableHttpResponse getWeatherData() {
+        try {
+            HttpGet getRequest = new HttpGet(new URIBuilder(WEATHER_ENDPOINT)
+                    .setParameter("id", CITY_ID)
+                    .setParameter("appid", API_KEY)
+                    .build());
+            CloseableHttpClient httpclient = HttpClients.createDefault();
```
Comment (breedloj) [inline samples/cityguide/src/com/amazon/ask/cityguide/handlers/GoOutIntentHandler.java]: Reuse this client.

File: samples/cityguide/src/com/amazon/ask/cityguide/handlers/GoOutIntentHandler.java [hunk 8]
```diff
@@ -0,0 +1,91 @@
...
+
+    @Override
+    public boolean canHandle(HandlerInput handlerInput, IntentRequest intentRequest) {
+        return intentRequest.getIntent().getName().equals("GoOutIntent");
+    }
+
+    @Override
+    public Optional<Response> handle(HandlerInput handlerInput, IntentRequest intentRequest) {
+        try {
+            JsonNode node = MAPPER.readTree(getWeatherData().getEntity().getContent());
```
Comment (breedloj) [inline samples/cityguide/src/com/amazon/ask/cityguide/handlers/GoOutIntentHandler.java]: You will want to use try-with-resources on the CloseableHttpResponse, or it'll never be closed and you'll run into connection starvation issues.

File: samples/cityguide/src/com/amazon/ask/cityguide/utils/City.java [hunk 9]
```diff
@@ -0,0 +1,34 @@
...
+
+     or in the "license" file accompanying this file. This file is distributed on an "AS IS" BASIS,
+     WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for
+     the specific language governing permissions and limitations under the License.
+*/
+
+package com.amazon.ask.cityguide.utils;
+
+public enum City {
+    GLOUCESTIRE("Gloucestire", "01930");
```
Comment (breedloj) [inline samples/cityguide/src/com/amazon/ask/cityguide/utils/City.java]: *Gloucester

File: samples/cityguide/README.md [hunk 10]
```diff
@@ -0,0 +1,147 @@
...
+## Concepts
+This sample shows how to create a Lambda function for handling Alexa Skill requests that:
+
+- Custom slot type: demonstrates using custom slot types to handle a finite set of known values
+- Dialog and Session state: Handles two models, both a one-shot ask and tell model, and a multi-turn dialog model.
+- Make API calls to other services.
+- Usage of ResourceBundle to handle internalization.
+
+## Setup
+To run this example skill you need to do two things. The first is to deploy the example code in lambda, and the second is to configure the Alexa skill to use Lambda.
```
Comment (breedloj) [inline samples/cityguide/README.md]: *Lambda

**Review by breedloj — COMMENTED** [review]
Port looks good overall. Mostly have minor cleanup comments.

File: samples/cityguide/src/com/amazon/ask/cityguide/POJO/CityInformation.java (hunk 4 above)
Comment (pbheemag) [reply samples/cityguide/src/com/amazon/ask/cityguide/POJO/CityInformation.java]: Used intellij feature of auto generating constructor. Will fix it.

File: samples/cityguide/README.md [hunk 11]
```diff
@@ -0,0 +1,147 @@
...
+      "intents": [
+        {
+          "name": "AboutIntent",
+          "samples": [
+            "about",
+            "tell me about this place"
+          ],
+          "slots": []
+        },
+
```
Comment (breedloj) [inline samples/cityguide/README.md]: The JSON in the docs still needs some whitespace cleanup.

**Review by breedloj — APPROVED** [review]
Nice work on this!

File: samples/cityguide/README.md (hunk 11 above)
Comment (pbheemag) [reply samples/cityguide/README.md]: fixed.

**Comment by pbheemag** [comment]
https://github.com/alexa/skill-sample-nodejs-city-guide

**Review by sungolivia — APPROVED** [review]

**Comment by pbheemag** [comment]
Closing this pull request since the changes have been moved to a separate repo. 
https://github.com/alexa/skill-sample-java-city-guide

**Pull request closed without merging**